drivers/amazonec2/amazonec2.go

```diff
@@ -1336,271 +1513,215 @@ func (d *Driver) configureSecurityGroups(groupNames []string) error {
-				IpProtocol: aws.String("udp"),
-				FromPort:   aws.Int64(vxlanPorts[0]),
-				ToPort:     aws.Int64(vxlanPorts[1]),
-				UserIdGroupPairs: []*ec2.UserIdGroupPair{
-					{
-						GroupId: group.GroupId,
-					},
-				},
-			})
+		// IPv6
+		if _, ok := hasV6Inbound[key]; !ok {
+			inboundPerms = append(inboundPerms,
+				newIpPermission(proto, from, to, nil,
+					[]*ec2.Ipv6Range{{CidrIpv6: aws.String(ipv6Range), Description: aws.String(description)}}))
 		}
+	}
 
-		// typha
-		if _, ok := hasPortsInbound[fmt.Sprintf("%d/tcp", typhaPorts[0])]; !ok {
-			inboundPerms = append(inboundPerms, &ec2.IpPermission{
-				IpProtocol: aws.String("tcp"),
-				FromPort:   aws.Int64(typhaPorts[0]),
-				ToPort:     aws.Int64(typhaPorts[1]),
-				UserIdGroupPairs: []*ec2.UserIdGroupPair{
-					{
-						GroupId: group.GroupId,
-					},
-				},
-			})
+	// Helper for group-based rules
+	addGroupRule := func(proto string, from, to int64) {
+		key := fmt.Sprintf("%d/%s", from, proto)
+		if _, ok := hasGroupRule[key]; !ok {
+			inboundPerms = append(inboundPerms,
+				newIpPermission(proto, from, to, nil, nil,
+					&ec2.UserIdGroupPair{GroupId: group.GroupId, Description: aws.String(description)}))
 		}
+	}
 
-		// flannel
-		if _, ok := hasPortsInbound[fmt.Sprintf("%d/udp", flannelPorts[0])]; !ok {
-			inboundPerms = append(inboundPerms, &ec2.IpPermission{
-				IpProtocol: aws.String("udp"),
-				FromPort:   aws.Int64(flannelPorts[0]),
-				ToPort:     aws.Int64(flannelPorts[1]),
-				UserIdGroupPairs: []*ec2.UserIdGroupPair{
-					{
-						GroupId: group.GroupId,
-					},
-				},
-			})
+	// Rules for rancher-nodes security group
+	if *group.GroupName == defaultSecurityGroup && hasTagKey(group.Tags, machineSecurityGroupName) {
+		// dual-stack ports
+		addDualStack("tcp", sshPort, sshPort)
+		addDualStack("tcp", dockerPort, dockerPort)
+		addDualStack("tcp", rancherWebhookPort, rancherWebhookPort)
+		addDualStack("tcp", kubeApiPort, kubeApiPort)
+		addDualStack("tcp", httpPort, httpPort)
+		addDualStack("tcp", httpsPort, httpsPort)
+		addDualStack("tcp", nodePorts[0], nodePorts[1])
+		addDualStack("udp", nodePorts[0], nodePorts[1])
+
+		// group-pair rules
+		addGroupRule("tcp", supervisorPort, supervisorPort)
+		addGroupRule("tcp", etcdPorts[0], etcdPorts[1])
+		addGroupRule("tcp", typhaPorts[0], typhaPorts[1])
+		addGroupRule("tcp", otherKubePorts[0], otherKubePorts[1])
+		addGroupRule("tcp", kubeProxyPorts[0], kubeProxyPorts[1])
+		addGroupRule("tcp", nodeExporter, nodeExporter)
+		addGroupRule("tcp", calicoPort, calicoPort)
+		addGroupRule("udp", vxlanPorts[0], vxlanPorts[1])
+		addGroupRule("udp", flannelPorts[0], flannelPorts[1])
+	}
+
+	// Custom open ports from configuration
+	for _, p := range d.OpenPorts {
+		port, protocol := driverutil.SplitPortProto(p)
+		portNum, err := strconv.ParseInt(port, 10, 0)
+		if err != nil {
+			return nil, fmt.Errorf("invalid port number %s: %s", port, err)
 		}
+		addDualStack(protocol, portNum, portNum)
```

Comment:
Does this mean there is no way to configure rules for just IPv4, or IPv6? I would think IPv6 exclusive clusters should only define rules for IPv6. Mostly just curious though, I'm not sure if that was a requirement in the RFD, if not then no worries.

Reply:
The current design is to include both IPv4 and IPv6 rules in the Rancher-managed security group rancher-nodes. 
This is because the same security group may be shared by all EC2 instances, whether they belong to single-stack or dual-stack clusters, when users do not specify a custom security group in their machine pools.

Reply:
Yeah that makes sense, I just saw that this particular call is made when handling `Custom open ports from configuration`. So this would also apply for security groups other than the default one, right? Probably not a big deal, but just wanted to understand if that is required for the IPv6 only configuration or just something we decided on - won't block on this.

Reply:
> So this would also apply for security groups other than the default one, right?
 
Correct. 

For the same reason mentioned in my previous message, we decided to add IPv6 rules for user-defined ports. Admittedly, this detail is not explicitly called out in the RFD, as it came up during implementation. The RFD is intended to capture high-level design decisions and does not necessarily include every implementation detail.